File: server/src/models/Ingest.js
import mongoose from 'mongoose';

const ingestSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  type: {
    type: String,
    enum: ['url', 'file', 'text', 'multiple_files', 'multiple_sources'],
    required: true
  },
  content: {
    type: String,
    required: true
  },
  originalContent: {
    type: String // Store original content before processing
  },
  metadata: {
    filename: String,
    mimeType: String,
    size: Number,
    url: String,
    title: String,
    description: String,
    tags: [String],
    language: String,
    encoding: String
  },
  status: {
    type: String,
    enum: ['processing', 'completed', 'failed'],
    default: 'processing'
  },
  processedContent: {
    type: String // Cleaned/processed version of content
  },
  extractedData: {
    type: Map,
    of: mongoose.Schema.Types.Mixed,
    default: new Map()
  },
  error: {
    message: String,
    stack: String,
    timestamp: Date
  },
  processingStats: {
    startTime: Date,
    endTime: Date,
    duration: Number,
    tokensProcessed: Number,
    chunksCreated: Number
  }
}, {
  timestamps: true
});

// Indexes for performance
ingestSchema.index({ userId: 1, createdAt: -1 });
ingestSchema.index({ type: 1 });
ingestSchema.index({ status: 1 });

// Method to mark as completed
ingestSchema.methods.markCompleted = function(processedContent, extractedData = {}) {
  this.status = 'completed';
  this.processedContent = processedContent;
  this.extractedData = new Map(Object.entries(extractedData));
  this.processingStats.endTime = new Date();
  this.processingStats.duration = this.processingStats.endTime - this.processingStats.startTime;
  return this.save();
};

// Method to mark as failed
ingestSchema.methods.markFailed = function(error) {
  this.status = 'failed';
  this.error = {
    message: error.message,
    stack: error.stack,
    timestamp: new Date()
  };
  this.processingStats.endTime = new Date();
  this.processingStats.duration = this.processingStats.endTime - this.processingStats.startTime;
  return this.save();
};

// Method to start processing
ingestSchema.methods.startProcessing = function() {
  this.status = 'processing';
  this.processingStats.startTime = new Date();
  return this.save();
};

export default mongoose.model('Ingest', ingestSchema);

File: server/src/models/Ingest.py
from mongoengine import Document, StringField, DictField, DateTimeField, MapField, IntField, ListField, EmbeddedDocument, EmbeddedDocumentField, FloatField
import datetime

class ErrorInfo(EmbeddedDocument):
    message = StringField()
    stack = StringField()
    timestamp = DateTimeField()

class ProcessingStats(EmbeddedDocument):
    startTime = DateTimeField()
    endTime = DateTimeField()
    duration = FloatField()
    tokensProcessed = IntField()
    chunksCreated = IntField()

class Ingest(Document):
    userId = StringField(required=True)
    type = StringField(choices=['url', 'file', 'text', 'multiple_files', 'multiple_sources'], required=True)
    content = StringField(required=True)
    originalContent = StringField()
    metadata = DictField()
    status = StringField(choices=['processing', 'completed', 'failed'], default='processing')
    processedContent = StringField()
    extractedData = MapField(field=StringField())
    error = EmbeddedDocumentField(ErrorInfo)
    processingStats = EmbeddedDocumentField(ProcessingStats)
    createdAt = DateTimeField(default=datetime.datetime.utcnow)
    updatedAt = DateTimeField(default=datetime.datetime.utcnow)

    meta = {
        'indexes': [
            {'fields': ['userId', '-createdAt']},
            {'fields': ['type']},
            {'fields': ['status']}
        ]
    }

    def mark_completed(self, processed_content, extracted_data=None):
        self.status = 'completed'
        self.processedContent = processed_content
        self.extractedData = extracted_data or {}
        if self.processingStats:
            self.processingStats.endTime = datetime.datetime.utcnow()
            self.processingStats.duration = (self.processingStats.endTime - self.processingStats.startTime).total_seconds() * 1000
        self.save()

    def mark_failed(self, error):
        self.status = 'failed'
        self.error = ErrorInfo(
            message=str(error),
            stack=getattr(error, 'stack', ''),
            timestamp=datetime.datetime.utcnow()
        )
        if self.processingStats:
            self.processingStats.endTime = datetime.datetime.utcnow()
            self.processingStats.duration = (self.processingStats.endTime - self.processingStats.startTime).total_seconds() * 1000
        self.save()

    def start_processing(self):
        self.status = 'processing'
        if not self.processingStats:
            self.processingStats = ProcessingStats()
        self.processingStats.startTime = datetime.datetime.utcnow()
        self.save()
